apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1350 @@
+        address challenger;                 // Address that challenged the action
+        uint64 endDate;                     // Last date the submitter can raise a dispute against the challenge
+        bytes context;                      // Link to a human-readable text providing context for the challenge
+        uint256 settlementOffer;            // Amount of collateral tokens the challenger would accept without involving the arbitrator
+        uint256 arbitratorFeeAmount;        // Amount of arbitration fees paid by the challenger in advance
+        ERC20 arbitratorFeeToken;           // ERC20 token used for the arbitration fees paid by the challenger in advance
+        ChallengeState state;               // Current state of the action challenge
+        bool submitterFinishedEvidence;     // Whether the action submitter has finished submitting evidence for a raised dispute
+        bool challengerFinishedEvidence;    // Whether the action challenger has finished submitting evidence for a raised dispute
+        uint256 disputeId;                  // Identification number of the dispute on the arbitrator
+        uint256 ruling;                     // Ruling given for the action's dispute
+    }
+
+    struct CollateralRequirement {
+        ERC20 token;                        // ERC20 token to be used for collateral
+        uint256 actionAmount;               // Amount of collateral token that will be locked from the submitter's staking pool every time an action is created
+        uint256 challengeAmount;            // Amount of collateral token that will be locked from the challenger's own balance every time an action is challenged
+        uint64 challengeDuration;           // Challenge duration in seconds, during which the submitter can raise a dispute
+        Staking staking;                    // Staking pool cache for the collateral token
+    }
+
+    struct DisputableInfo {
+        bool registered;                                                    // Whether a Disputable app is registered
+        uint256 collateralRequirementsLength;                               // Identification number of the next collateral requirement instance
+        mapping (uint256 => CollateralRequirement) collateralRequirements;  // List of collateral requirements indexed by id
+    }
+
+    StakingFactory public stakingFactory;                           // Staking factory to be used for the collateral staking pools
+
+    uint256 private settingsLength;
+    mapping (uint256 => Setting) private settings;                  // List of historic settings indexed by ID
+    mapping (address => uint256) private lastSettingSignedBy;       // List of last setting signed by user
+
+    uint256 private actionsLength;
+    mapping (uint256 => Action) private actions;                    // List of actions indexed by ID
+    mapping (address => DisputableInfo) private disputableInfos;    // Mapping of disputable address => disputable infos
```

Comment:
Nitpick, but perhaps we should also move `disputableInfos` to be right after the settings so we consistently order our concepts as:

1. Signers
1. Settings
1. Disputables
1. Actions